dataplicity/m2m/commandservice.py

```diff
@@ -18,102 +18,108 @@
 
 from lomond.errors import WebSocketError
 
-from .. constants import CHUNK_SIZE
+from ..limiter import LimitReached
+from ..constants import CHUNK_SIZE
 
 
-log = logging.getLogger('m2m')
+log = logging.getLogger("m2m")
 
 
 class CommandService(threading.Thread):
     """Runs a command and sends the stdout over m2m."""
 
-    def __init__(self, channel, command):
+    def __init__(self, limiter, channel, command):
+        self.limiter = limiter
         self._repr = "CommandService({!r}, {!r})".format(channel, command)
-        super(CommandService, self).__init__(
-            args=(channel, command),
-            target=self.run_service
-        )
-        self.start()
+        try:
+            with limiter():
+                super(CommandService, self).__init__(
+                    args=(channel, command), target=self.run_service
+                )
+                self.start()
+        except Exception as error:
+            log.warning("unable to launch %r; %s", self, error)
+            self.send_error(channel, "error", str(error))
+            channel.close()
 
     def __repr__(self):
         return self._repr
 
     @classmethod
     def send_error(cls, channel, status, msg, **extra):
         """Send a control packet with an error"""
-        error = {
-            "service": "command",
-            "type": "error",
-            "status": status,
-            "msg": msg
-        }
+        error = {"service": "command", "type": "error", "status": status, "msg": msg}
         error.update(extra)
         channel.send_control(error)
 
     def run_service(self, channel, command):
         """Run the thread and log exceptions."""
         try:
-            self._run_service(channel, command)
-        except Exception:
-            log.exception("error running %r", self)
-            self.send_error(channel, "error", "error running command")
+            try:
+                self._run_service(channel, command)
+            except Exception:
+                log.exception("error running %r", self)
+                self.send_error(channel, "error", "error running command")
+        finally:
+            self.limiter.decrement()
 
     def _run_service(self, channel, command):
         """Run command and send stdout over m2m."""
         log.debug("%r started", self)
         process = subprocess.Popen(
-            command,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-            shell=True
+            command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True
         )
         bytes_sent = 0
         stdout_fh = process.stdout.fileno()
         stderr_fh = process.stderr.fileno()
+
```

Comment:
I missed this in an earlier updating when I was replacing select with poll.

This needs testing as I think it is only invoked from the mobile app.